limiter.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package rate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// Limiter is used to determine if a request for a given resource and action
+// should be allowed.
+// TODO: expand this doc
+type Limiter struct {
+	limits map[string]*Limit
+
+	sync.RWMutex
+
+	store store
+}
+
+// NewLimiter will create a Limiter with the provided limits and max size. The
+// limits must each be unique, where uniqueness is determined by the
+// combination of "resource", "action", and "per". The maxSize must be greater
+// than zero. This size is the number of individual quotas that can be stored
+// in memory at any given time. Once this size is reached, requests that would
+// result in a new quota being inserted will not be allowed. Requests that
+// correspond to existing quotas will still be processed as normal. Space will
+// become available once quotas expire and are removed.
+//
+// Supported options are:
+//   - WithNumBuckets: Sets the number of buckets used for expiring quotas.
+//     This must be greater than zero, and defaults to DefaultNumBuckets. A
+//     larger number of buckets can increase the efficiency at which expired
+//     quotas are deleted to free up space. However, it does also marginally
+//     increase the amount of memory needed, and can increase the frequency
+//     in which the delete routine runs and must acquire a lock.
+func NewLimiter(ctx context.Context, limits []*Limit, maxSize int, o ...Option) (*Limiter, error) {
```

Comment:
Can you add how `ctx` is used to the go docs?

Reply:
I eliminated the need for this ctx. Instead any context usage is internalized in the expireableStore, with the `shutdown` method being the mechanism to stop the delete goroutine and render the expireableStore, and therefore the Limiter unusable. I also added a `Shutdown` method to the Limiter. The idea being that `Shutdown` should be used when:
- As part of clean shutdown of a system using a Limiter
- If a system wishes to replace a Limiter due to changing its limits, like via a SIGHUP, the previous Limiter should be shutdown so it can be disposed of.

I plan to revisit this to include more details documentation around how to use a Limiter.